sdks/python/apache_beam/io/vcfio.py

```diff
@@ -0,0 +1,378 @@
+        specified, this field will be set to `*`.
+      info (dict): A map of additional variant call information. The key is
+        specified in the VCF record and the type of the value is specified by
+        the VCF header FORMAT.
+    """
+    self.name = name
+    self.genotype = genotype or []
+    self.phaseset = phaseset
+    self.info = info or {}
+
+  def __eq__(self, other):
+    return (self.name == other.name and
+            self.genotype == other.genotype and
+            self.phaseset == other.phaseset and
+            self.info == other.info)
+
+  def __repr__(self):
+    return ', '.join(
+        [str(s) for s in [self.name, self.genotype, self.phaseset, self.info]])
+
+
+class _VcfSource(filebasedsource.FileBasedSource):
+  r"""A source for reading VCF files.
+
+  Parses VCF files (version 4) using PyVCF library. If file_pattern specifies
+  multiple files, then the header from each file is used separately to parse
+  the content. However, the output will be a uniform PCollection of
+  :class:`Variant` objects.
+  """
+
+  DEFAULT_VCF_READ_BUFFER_SIZE = 65536  # 64kB
+
+  def __init__(self,
+               file_pattern,
+               min_bundle_size=0,
+               compression_type=CompressionTypes.AUTO,
+               buffer_size=DEFAULT_VCF_READ_BUFFER_SIZE,
+               validate=True):
+    super(_VcfSource, self).__init__(file_pattern, min_bundle_size,
+                                     compression_type=compression_type,
+                                     validate=validate)
+
+    self._coder = coders.StrUtf8Coder() # VCF files are UTF-8 per schema.
+    self._header_lines_per_file = {}
+    self._text_source = TextSource(
+        file_pattern,
+        min_bundle_size,
+        compression_type,
+        strip_trailing_newlines=True,
+        coder=self._coder,
+        buffer_size=buffer_size,
+        validate=validate,
+        skip_header_lines=0,
+        header_matcher_predicate=lambda x: x.startswith('#'))
+
+  def _process_header_lines_matching_predicate(self, file_name, header_lines):
+    self._header_lines_per_file[file_name] = header_lines
+
+  def read_records(self, file_name, range_tracker):
+    def line_generator():
+      header_processed = False
+      for line in self._text_source.read_records(
+          file_name,
+          range_tracker,
+          self._process_header_lines_matching_predicate):
+        if not header_processed and file_name in self._header_lines_per_file:
+          for header in self._header_lines_per_file[file_name]:
+            yield header
+          header_processed = True
+        # PyVCF has explicit str() calls when parsing INFO fields, which fails
+        # with UTF-8 decoded strings. Encode the line back to UTF-8.
+        yield self._coder.encode(line)
+    try:
+      vcf_reader = vcf.Reader(fsock=line_generator())
+    except SyntaxError as e:
+      raise ValueError('Invalid VCF header: %s' % str(e))
+    while True:
```

Comment:
<!--thread_id:cc_144384975_t; commit:6587a5a34ad6aff2567df759a830067647adbaa5; resolved:0-->
<!--section:context-quote-->
> **jkff** wrote:
> I suspect that, without the exception handling below, this loop could have been written as a one-liner generator expression

<!--section:body-->
+1. Please simplify by moving exception handling to a single try block. I expect vcf.Reader() invocation to not expand line_generator() here anyways (if it does it will be very inefficient and we could run into memory issues).